File: pipeline/models/traits.py
"""Pydantic model for traits.json."""

from __future__ import annotations

from .common import LenientModel


class Trait(LenientModel):
    """An NPC trait with optional parameters."""

    id: str
    name: str
    parameterized: bool
    effect: str
    derivedEffects: dict[str, str]
    # Optional — only present when parameterized=True
    paramLabel: str | None = None
    paramType: str | None = None

File: src/lib/dtd/schemas/traits.ts
/**
 * Zod schema for traits data — replaces former Pydantic model.
 *
 * Top-level JSON is a bare array of Trait objects.
 */

import { z } from "zod";

// ---------------------------------------------------------------------------
// Trait (lenient — extra fields allowed)
// ---------------------------------------------------------------------------

export const Trait = z
	.object({
		id: z.string(),
		name: z.string(),
		parameterized: z.boolean(),
		effect: z.string(),
		derivedEffects: z.record(z.string(), z.string()),
		paramLabel: z.string().nullish(),
		paramType: z.string().nullish(),
	})
	.passthrough();
export type Trait = z.infer<typeof Trait>;

// ---------------------------------------------------------------------------
// File-level schema (bare array)
// ---------------------------------------------------------------------------

export const TraitsFile = z.array(Trait);
export type TraitsFile = z.infer<typeof TraitsFile>;
